crates/uv/tests/lock.rs

```diff
@@ -9918,3 +9918,299 @@ fn lock_exclude_unnecessary_python_forks() -> Result<()> {
 
     Ok(())
 }
+
+/// Lock a requirement from PyPI.
+#[test]
+fn lock_constrained_environment() -> Result<()> {
+    let context = TestContext::new("3.12");
+
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "project"
+        version = "0.1.0"
+        requires-python = ">=3.12"
+        dependencies = ["black"]
+
+        [tool.uv]
+        environments = ["platform_system != 'Windows'"]
+        "#,
+    )?;
+
+    uv_snapshot!(context.filters(), context.lock(), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+
+    ----- stderr -----
+    warning: `uv lock` is experimental and may change without warning
+    Resolved 7 packages in [TIME]
+    "###);
+
+    let lock = fs_err::read_to_string(context.temp_dir.join("uv.lock")).unwrap();
+
+    // Because we're _not_ locking for Windows, `colorama` should not be included.
+    insta::with_settings!({
+        filters => context.filters(),
+    }, {
+        assert_snapshot!(
+            lock, @r###"
+        version = 1
+        requires-python = ">=3.12"
+        environment-markers = [
+            "platform_system != 'Windows'",
+        ]
+        supported-markers = [
+            "platform_system != 'Windows'",
+        ]
+
+        [options]
+        exclude-newer = "2024-03-25 00:00:00 UTC"
+
+        [[package]]
+        name = "black"
+        version = "24.3.0"
+        source = { registry = "https://pypi.org/simple" }
+        dependencies = [
+            { name = "click", marker = "platform_system != 'Windows'" },
+            { name = "mypy-extensions", marker = "platform_system != 'Windows'" },
+            { name = "packaging", marker = "platform_system != 'Windows'" },
+            { name = "pathspec", marker = "platform_system != 'Windows'" },
+            { name = "platformdirs", marker = "platform_system != 'Windows'" },
```

Comment:
I think there's another solution where we don't treat the `supported-markers` as the input forks, and instead just test for disjointness everywhere (which would avoid having to write these markers on dependencies). But this is much simpler to implement, and we can always change the strategy later.